Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,311 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+The finite Coxeter groups are classified as the four infinite families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter System.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ) : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relation terms corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
+  Function.uncurry fun i j =>
+    (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The relations corresponding to a Coxeter matrix. -/
+def toSet : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix.
+
+Note that it is defined for any matrix of natural numbers. Its value for non-Coxeter
+matrices should be regarded as junk. `IsCoxeterGroup` checks that the matrix `IsCoxeter`. -/
+abbrev Matrix.CoxeterGroup := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+/-- A Coxeter System `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem (W : Type*) [Group W] where
```

Comment:
After this definition, you should define `Matrix.coxeterSystemCoxeterGroup` which would say that there is a Coxeter system of type `M` on the group `M.CoxeterGroup` (with the identity isomorphism).